src/Common/src/System/Security/Cryptography/RSACng.EncryptDecrypt.cs

```diff
@@ -18,76 +18,102 @@ internal static partial class RSAImplementation
+
+        /// <summary>Encrypts data using the public key.</summary>
+        public override bool TryEncrypt(ReadOnlySpan<byte> source, Span<byte> destination, RSAEncryptionPadding padding, out int bytesWritten) =>
+            TryEncryptOrDecrypt(source, destination, padding, encrypt: true, bytesWritten: out bytesWritten);
+
+        /// <summary>Decrypts data using the private key.</summary>
+        public override bool TryDecrypt(ReadOnlySpan<byte> source, Span<byte> destination, RSAEncryptionPadding padding, out int bytesWritten) =>
+            TryEncryptOrDecrypt(source, destination, padding, encrypt: false, bytesWritten: out bytesWritten);
+
+        // Conveniently, Encrypt() and Decrypt() are identical save for the actual P/Invoke call to CNG. Thus, both
+        // array-based APIs invoke this common helper with the "encrypt" parameter determining whether encryption or decryption is done.
+        private unsafe byte[] EncryptOrDecrypt(byte[] data, RSAEncryptionPadding padding, bool encrypt)
         {
-            unsafe
+            if (data == null)
             {
-                return EncryptOrDecrypt(data, padding, Interop.NCrypt.NCryptEncrypt);
+                throw new ArgumentNullException(nameof(data));
+            }
+            if (padding == null)
+            {
+                throw new ArgumentNullException(nameof(padding));
             }
-        }
 
-        /// <summary>
-        ///     Decrypts data using the private key.
-        /// </summary>
-        public override byte[] Decrypt(byte[] data, RSAEncryptionPadding padding)
-        {
-            unsafe
+            using (SafeNCryptKeyHandle keyHandle = GetDuplicatedKeyHandle())
             {
-                return EncryptOrDecrypt(data, padding, Interop.NCrypt.NCryptDecrypt);
+                switch (padding.Mode)
+                {
+                    case RSAEncryptionPaddingMode.Pkcs1:
+                        return EncryptOrDecrypt(keyHandle, data, AsymmetricPaddingMode.NCRYPT_PAD_PKCS1_FLAG, null, encrypt);
+
+                    case RSAEncryptionPaddingMode.Oaep:
+                        IntPtr namePtr = Marshal.StringToHGlobalUni(padding.OaepHashAlgorithm.Name);
+                        try
+                        {
+                            var paddingInfo = new BCRYPT_OAEP_PADDING_INFO()
+                            {
+                                pszAlgId = namePtr,
+
+                                // It would nice to put randomized data here but RSAEncryptionPadding does not at this point provide support for this.
+                                pbLabel = IntPtr.Zero,
+                                cbLabel = 0,
+                            };
+                            return EncryptOrDecrypt(keyHandle, data, AsymmetricPaddingMode.NCRYPT_PAD_OAEP_FLAG, &paddingInfo, encrypt);
+                        }
+                        finally { Marshal.FreeHGlobal(namePtr); }
```

Comment:
It's not a style I've seen anywhere in corefx, and it stands out.

It also creates needless differences if we add a second thing to the finally.